In [ ]:
#Importing the libraries

import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split 
from sklearn import metrics

In [19]:
#Calling The Data

df = pd.read_csv('https://raw.githubusercontent.com/AdiPersonalWorks/Random/master/student_scores%20-%20student_scores.csv')

In [23]:
df

Hours  Scores
0     2.0      21
1     5.0      47
2     3.0      27
3     8.0      75
4     4.0      30
5     2.0      20
6     9.0      88
7     6.0      60
8     8.0      81
9     3.0      25
10    8.0      85
11    6.0      62
12    4.0      41
13    3.0      42
14    1.0      17
15    9.0      95
16    2.0      30
17    2.0      24
18    6.0      67
19    7.0      69
20    3.0      30
21    5.0      54
22    4.0      35
23    7.0      76
24    8.0      86

In [9]:
#Describing the Data
df.describe()

Hours     Scores
count  25.000000  25.000000
mean    5.012000  51.480000
std     2.525094  25.286887
min     1.100000  17.000000
25%     2.700000  30.000000
50%     4.800000  47.000000
75%     7.400000  75.000000
max     9.200000  95.000000

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Hours   25 non-null     float64
 1   Scores  25 non-null     int64  
dtypes: float64(1), int64(1)
memory usage: 528.0 bytes


## Key Observations

1. Mean Score of the population is 51.48
2. Mean hours of studying is 5
3. Here hours is of float data type and score is of integer data type

In [22]:
#Rounding off data

df['Hours'] = round(df['Hours'])

In [24]:
#Changing hours from float data type to integer data type

df['Hours'] = df['Hours'].astype('int64')

In [27]:
#Changing column names to lowercase

df.columns = df.columns.str.lower()

## Fitting a Regression Model

In [29]:
# Selecting the Feature or independent variable

feature_name = ['hours']

X = df[feature_name]

# Indepenent Variable

y = df.scores


In [30]:
#linear Regression Model

linreg = LinearRegression()

# Fitting the data
linreg.fit(X,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [31]:
print('Intercept:', linreg.intercept_ )
print('Coefficient:', linreg.coef_)

Intercept: 1.9799999999999827
Coefficient: [9.9]


# Solution:

In [39]:
# For 9.25 hours of study, marks scored

marks = linreg.intercept_ + 9.25*linreg.coef_

print(marks)

[93.555]


### Hence, for 9.25 hours of study a student will score approximately 93.5 marks

# Predictive Analysis:

In [33]:
#Checking the models prediction and generalisation

y_pred = linreg.predict(X)

In [34]:
# Checking performance using different metrics

print("RMSE:", np.sqrt(metrics.mean_squared_error(y,y_pred)))

print("r-Squared" , metrics.r2_score(y,y_pred))

RMSE: 5.0783461874905695
r-Squared 0.9579871030297975


In [35]:
df['y_pred'] = y_pred

In [36]:
df

hours  scores  y_pred
0       2      21   21.78
1       5      47   51.48
2       3      27   31.68
3       8      75   81.18
4       4      30   41.58
5       2      20   21.78
6       9      88   91.08
7       6      60   61.38
8       8      81   81.18
9       3      25   31.68
10      8      85   81.18
11      6      62   61.38
12      4      41   41.58
13      3      42   31.68
14      1      17   11.88
15      9      95   91.08
16      2      30   21.78
17      2      24   21.78
18      6      67   61.38
19      7      69   71.28
20      3      30   31.68
21      5      54   51.48
22      4      35   41.58
23      7      76   71.28
24      8      86   81.18